### PR TITLE
Created Ride Route

### PR DESCRIPTION
**Pull request by TrojanHorse0918**
Issues fixed in this PR: #2 

API Endpoint: `/api/team/create`
Methods allowed: `POST`

### Request Sample:
`http://127.0.0.1:3000/createRide`

Request Body:
```
{
    "host_user_id": 5,
    "start_location": "Mumbai",
    "end_location": "Pune",
    "start_time": "2024-12-31T23:00:00Z",
    "total_seats": 5,
    "booked_seats": 2,
    "total_price": 1000
}
```

### Response Sample:
```
{
    "ID": 0,
    "CreatedAt": "2023-12-19T06:15:23.3963644+05:30",
    "UpdatedAt": "2023-12-19T06:15:23.3963644+05:30",
    "DeletedAt": null,
    "id": 6,
    "host_user_id": 5,
    "host_user": {
        "ID": 0,
        "CreatedAt": "2023-12-19T06:14:44.920814+05:30",
        "UpdatedAt": "2023-12-19T06:14:44.920814+05:30",
        "DeletedAt": null,
        "id": 5,
        "name": "Manan Shah",
        "email": "manan04efbfgshah@gmail.com",
        "profile_picture_url": "https://www.google.com"
    },
    "start_location": "Mumbai",
    "end_location": "Pune",
    "start_time": "2024-12-31T23:00:00Z",
    "total_seats": 5,
    "booked_seats": 2,
    "total_price": 1000
}
```

###Error Cases

If using any other method except `POST`
Status: 400
```
{
    message: "Request method is not POST"
}
```

If there is an error parsing the JSON
Status: 400
```
{
    message: "Error parsing JSON"
}
```

If the request body isn't correctly formatted as per the ride model
Status: 400
```
{
    message: "Error validating ride"
}
```

If the host user does not exist
Status: 400
```
{
    message: "Host user does not exist"
}
```

If the start time of the ride is before current time
Status: 400
```
{
    message: "Start time is in the past"
}
```

If the total number of seats in the ride is lesser than the number of seats booked
Status: 400
```
{
    message: "Total seats available is less than booked seats"
}
```

Internal error
Status: 502
```
{
    message: "Error creating ride"
}
```

### CODE REVIEW

File: routes/ride/createRide.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+	//Logs a 400 error if the Ride struct is invalid
+	err = ValidateRide(ride) // Validate the Ride struct
+	if err != nil {
+		log.Printf("Error validating ride: %v\n", err)
+		return c.Status(400).SendString("Error validating ride")
+	}
+
+	//DATA VALIDATION
+	//Checks if the host user ID exists
+	var hostUser []models.User
```
Comment (RohanKhatua) [inline routes/ride/createRide.go:54]: why is this an array?

File: routes/ride/createRide.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+	result := database.Database.Db.Where("id = ?", ride.HostUserID).Find(&hostUser)
```
Comment (RohanKhatua) [inline routes/ride/createRide.go:55]: Host IDs are unique, just Db.First will most likely work

File: routes/ride/createRide.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+	if result.Error != nil {
+		log.Printf("Error finding host user: %v\n", result.Error)
+		return c.Status(502).SendString("Error finding host user")
+	}
+	if len(hostUser) == 0 {
```
Comment (RohanKhatua) [inline routes/ride/createRide.go:60]: result := db.Find(&users)
if errors.Is(result.Error, gorm.ErrRecordNotFound) {
    // No records found
} else if result.Error != nil {
    // Some other error occurred
} else {
    // Records found
}

Better way to check if no records were found using Find

File: routes/ride/createRide.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+	}
+
+	//Checks if start time is in the future
+	if ride.StartTime.Before(time.Now()) {
+		log.Printf("Start time is in the past")
+		return c.Status(400).SendString("Start time is in the past")
+	}
+
+	//Checks if the host user has enough seats
+	if ride.TotalSeats < ride.BookedSeats {
```
Comment (RohanKhatua) [inline routes/ride/createRide.go:72]: Should also disallow if booked seats = total seats, which implies car is full

File: routes/ride/createRide.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+	//Obtaining the user from the db using the host user ID
+	ride.HostUser = hostUser[0]
+
+	//CREATING THE RIDE
+	result = database.Database.Db.Create(&ride)
+
+	if result.Error != nil {
+		log.Printf("Error creating ride: %v\n", result.Error)
+		return c.Status(502).SendString("Error creating ride")
```
Comment (RohanKhatua) [inline routes/ride/createRide.go:85]: Why 502? This is not necessarily a gateway issue.

**Review by RohanKhatua — CHANGES_REQUESTED** [review]
Please check review comments

File: routes/ride/createRide.go [hunk 6]
```diff
@@ -32,7 +33,7 @@ func CreateRide(c *fiber.Ctx) error {
 	//Logs a 400 error if the request method is not POST
 	if c.Method() != "POST" {
```
Comment (RohanKhatua) [inline routes/ride/createRide.go:34]: Why check this? It is going to be registered as a POST route only.

**Review by TrojanHorse0918 — COMMENTED** [review]
The requested changes have been made, have used db.First() instead of db.Find() for user ID checking as db.Find() did not raise a gormError when no users were found.

File: routes/ride/createRide.go [hunk 7]
```diff
@@ -69,20 +69,20 @@ func CreateRide(c *fiber.Ctx) error {
...
-		log.Printf("Total seats available is less than booked seats")
-		return c.Status(400).SendString("Total seats available is less than booked seats")
+	if ride.TotalSeats <= ride.BookedSeats {
+		log.Printf("Total seats available should be more than booked seats")
+		return c.Status(400).SendString("Total seats available should be more than booked seats")
 	}
 
 	//Obtaining the user from the db using the host user ID
-	ride.HostUser = hostUser[0]
+	ride.HostUser = hostUser
```
Comment (RohanKhatua) [inline routes/ride/createRide.go:78]: That is an FKey Reference, and does not need to be set explicitly.

**Review by RohanKhatua — CHANGES_REQUESTED** [review]

**Review by RohanKhatua — APPROVED** [review]

**Comment by RohanKhatua** [comment]
merging for time

**Pull request merged**